File: wlt/web.py
import ipaddress
import logging
import os
import socket

from flask import Flask, jsonify, render_template, request

from .config import load_config
from .nft import NftHandler, get_duration_label, get_group_selection

logging.basicConfig(level=logging.INFO)

CONFIG = load_config()
nft = NftHandler(CONFIG.nftables)

# Gunicorn can load this module as a config source via `-c python:wlt.web`.
_host = CONFIG.flask.host
bind = f"[{_host}]:{CONFIG.flask.port}" if ":" in _host else f"{_host}:{CONFIG.flask.port}"

app = Flask(__name__, template_folder=os.path.join(os.path.dirname(os.path.dirname(__file__)), "templates"))
logger = app.logger


@app.after_request
def _cors(resp):
    # The SPA loads from one split-horizon host and fetches the sibling family's
    # API cross-origin. Allow any *.gaof.net origin (no credentials are used).
    origin = request.headers.get("Origin")
    if origin:
        try:
            host = origin.split("//", 1)[1].split("/", 1)[0].split(":", 1)[0]
        except IndexError:
            host = ""
        if host == "gaof.net" or host.endswith(".gaof.net"):
            resp.headers["Access-Control-Allow-Origin"] = origin
            resp.headers["Vary"] = "Origin"
    return resp


def _normalize_ip(raw: str) -> str:
    try:
        addr = ipaddress.ip_address(raw)
    except ValueError:
        return raw
    if isinstance(addr, ipaddress.IPv6Address) and addr.ipv4_mapped:
        return str(addr.ipv4_mapped)
    return str(addr)


def get_client_info() -> tuple[str, str]:
    ip = _normalize_ip(request.remote_addr or "127.0.0.1")
    try:
        hostname = socket.gethostbyaddr(ip)[0]
    except (socket.herror, OSError):
        hostname = ip
    return ip, hostname


def _family(ip: str) -> int:
    try:
        return ipaddress.ip_address(ip).version
    except ValueError:
        return 4


def _status_payload() -> dict:
    ip, hostname = get_client_info()
    family = _family(ip)
    map_name = CONFIG.map_for(family)
    if not map_name:
        return {
            "ip": ip,
            "hostname": hostname,
            "family": family,
            "available": False,
            "groups": [],
            "current_outlet": "默认",
            "expires": None,
            "time_limits": [
                {"label": get_duration_label(t), "value": t} for t in CONFIG.time_limits
            ],
        }

    entry = nft.get_entry(ip, map_name)
    mark_value = entry.mark if entry else None

    groups = []
    current_labels = []
    for idx, group in enumerate(CONFIG.outlet_groups):
        outlets = group.outlets_for(family)
        if not outlets:
            continue
        selection = (
            get_group_selection(mark_value, group.mask, outlets)
            if mark_value is not None
            else None
        )
        if selection:
            current_labels.append(selection)
        groups.append(
            {
                "title": group.title,
                "field": f"group_{idx}",
                "options": list(outlets.keys()),
                "selected": selection or next(iter(outlets.keys())),
            }
        )

    current_outlet = "默认"
    if mark_value is not None:
        current_outlet = " + ".join(current_labels) if current_labels else hex(mark_value)

    return {
        "ip": ip,
        "hostname": hostname,
        "family": family,
        "available": True,
        "groups": groups,
        "current_outlet": current_outlet,
        "expires": entry.expires if entry else None,
        "time_limits": [
            {"label": get_duration_label(t), "value": t} for t in CONFIG.time_limits
        ],
    }


@app.route("/", methods=["GET"])
def index():
    portal = CONFIG.portal
    return render_template(
        "spa.html.j2",
        v4_host=portal.v4_host,
        v6_host=portal.v6_host,
    )


@app.route("/api/status", methods=["GET"])
def api_status():
    return jsonify(_status_payload())


@app.route("/api/open", methods=["POST"])
def api_open():
    ip, _ = get_client_info()
    family = _family(ip)
    map_name = CONFIG.map_for(family)
    if not map_name:
        return jsonify({"ok": False, "message": "当前协议族暂不支持出口选择", "family": family}), 400

    mark_value = 0
    selected_labels = []
    for idx, group in enumerate(CONFIG.outlet_groups):
        outlets = group.outlets_for(family)
        if not outlets:
            continue
        outlet = request.form.get(f"group_{idx}")
        if not outlet or outlet not in outlets:
            return jsonify({"ok": False, "message": f"无效的出口选择：{group.title}", "family": family}), 400
        selected_labels.append(outlet)
        mark_value |= outlets[outlet] & group.mask

    try:
        hours = int(request.form.get("hours", ""))
        if hours not in CONFIG.time_limits:
            raise ValueError
    except ValueError:
        return jsonify({"ok": False, "message": "无效的时限选择", "family": family}), 400

    nft.delete_element(ip, map_name)
    if nft.add_element(ip, hex(mark_value), hours, map_name):
        duration = f"{hours}小时" if hours else "永久"
        message = f"IPv{family} 已开通：「{' + '.join(selected_labels)}」，{duration}"
        return jsonify({"ok": True, "message": message, "family": family})
    return jsonify({"ok": False, "message": "设置网络出口失败", "family": family}), 500


@app.route("/api/close", methods=["POST"])
def api_close():
    ip, _ = get_client_info()
    family = _family(ip)
    map_name = CONFIG.map_for(family)
    if map_name and nft.delete_element(ip, map_name):
        return jsonify({"ok": True, "message": f"IPv{family} 已重置", "family": family})
    return jsonify({"ok": False, "message": "重置网络失败", "family": family}), 500


def main():
    app.run(host=CONFIG.flask.host, port=CONFIG.flask.port, debug=CONFIG.flask.debug)

File: src/web.rs
//! Web portal: static SPA page plus the JSON API consumed by it.

use std::collections::HashMap;
use std::net::SocketAddr;

use anyhow::{Context, Result};
use askama::Template;
use axum::Router;
use axum::extract::{ConnectInfo, Form, State};
use axum::http::{HeaderValue, StatusCode, header};
use axum::response::{Html, IntoResponse};
use axum::routing::{get, post};
use axum_server::tls_rustls::RustlsConfig;
use serde::Serialize;
use tokio::task::JoinSet;
use tower_http::cors::{AllowOrigin, CorsLayer};

use crate::app::{AppState, ip_family, normalize_ip};
use crate::config::{PortalConfig, WebConfig, duration_label};

#[derive(Template)]
#[template(path = "index.html")]
struct IndexTemplate {
    v4_host: String,
    v6_host: String,
    api_hosts_json: String,
}

#[derive(Serialize)]
struct TimeLimitPayload {
    label: String,
    value: u32,
}

#[derive(Serialize)]
struct GroupPayload {
    title: String,
    field: String,
    options: Vec<String>,
    selected: String,
}

#[derive(Serialize)]
struct StatusPayload {
    ip: String,
    hostname: Option<String>,
    family: u8,
    available: bool,
    groups: Vec<GroupPayload>,
    current_outlet: String,
    expires: Option<u32>,
    time_limits: Vec<TimeLimitPayload>,
}

#[derive(Serialize)]
struct ActionResponse {
    ok: bool,
    message: String,
    family: u8,
}

fn action(
    status: StatusCode,
    ok: bool,
    message: impl Into<String>,
    family: u8,
) -> impl IntoResponse {
    (
        status,
        axum::Json(ActionResponse {
            ok,
            message: message.into(),
            family,
        }),
    )
}

async fn index(State(state): State<AppState>) -> impl IntoResponse {
    let portal = &state.cfg.portal;
    let template = IndexTemplate {
        v4_host: portal.v4_host.clone().unwrap_or_default(),
        v6_host: portal.v6_host.clone().unwrap_or_default(),
        api_hosts_json: api_hosts_json(portal),
    };
    match template.render() {
        Ok(html) => Html(html).into_response(),
        Err(e) => {
            tracing::error!("template render failed: {e}");
            StatusCode::INTERNAL_SERVER_ERROR.into_response()
        }
    }
}

fn api_hosts_json(portal: &PortalConfig) -> String {
    let mut hosts = serde_json::Map::new();
    for (page_host, api_hosts) in &portal.hosts {
        let v4_host = api_hosts
            .v4_host
            .as_deref()
            .or(portal.v4_host.as_deref())
            .unwrap_or("");
        let v6_host = api_hosts
            .v6_host
            .as_deref()
            .or(portal.v6_host.as_deref())
            .unwrap_or("");
        hosts.insert(
            page_host.clone(),
            serde_json::json!({
                "4": v4_host,
                "6": v6_host,
            }),
        );
    }
    serde_json::Value::Object(hosts).to_string()
}

fn time_limits(state: &AppState) -> Vec<TimeLimitPayload> {
    state
        .cfg
        .time_limits
        .iter()
        .map(|&hours| TimeLimitPayload {
            label: duration_label(hours),
            value: hours,
        })
        .collect()
}

async fn api_status(
    State(state): State<AppState>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
) -> impl IntoResponse {
    let ip = normalize_ip(addr.ip());
    let family = ip_family(ip);
    let mut payload = StatusPayload {
        ip: ip.to_string(),
        hostname: state.resolve_hostname(ip).await,
        family,
        available: false,
        groups: vec![],
        current_outlet: "默认".into(),
        expires: None,
        time_limits: time_limits(&state),
    };
    let Some(map_name) = state.cfg.map_for(family) else {
        return axum::Json(payload);
    };
    payload.available = true;

    let entry = state
        .nft
        .get_entry(&payload.ip, map_name)
        .await
        .unwrap_or_else(|e| {
            tracing::error!("failed to fetch nftables entry for {ip}: {e:#}");
            None
        });
    let mark_value = entry.as_ref().map(|e| e.mark);
    payload.expires = entry.as_ref().and_then(|e| e.expires);

    let mut current_labels: Vec<&str> = Vec::new();
    for (idx, group) in state.cfg.outlet_groups.iter().enumerate() {
        if group.outlets_for(family).is_empty() {
            continue;
        }
        let selection = mark_value.and_then(|mark| group.selection_for(mark, family));
        if let Some(name) = selection {
            current_labels.push(name);
        }
        let display = group.display_outlets_for(family);
        payload.groups.push(GroupPayload {
            title: group.title.clone(),
            field: format!("group_{idx}"),
            options: display.iter().map(|&(name, _)| name.to_owned()).collect(),
            selected: selection.unwrap_or(display[0].0).to_owned(),
        });
    }
    payload.current_outlet = match mark_value {
        None => "默认".into(),
        Some(mark) if current_labels.is_empty() => format!("{mark:#x}"),
        Some(_) => current_labels.join(" + "),
    };
    axum::Json(payload)
}

async fn api_open(
    State(state): State<AppState>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
    Form(form): Form<HashMap<String, String>>,
) -> impl IntoResponse {
    let ip = normalize_ip(addr.ip());
    let family = ip_family(ip);
    let Some(map_name) = state.cfg.map_for(family) else {
        return action(
            StatusCode::BAD_REQUEST,
            false,
            "当前协议族暂不支持出口选择".into(),
            family,
        );
    };

    let mut mark_value: u32 = 0;
    let mut selected_labels: Vec<&str> = Vec::new();
    for (idx, group) in state.cfg.outlet_groups.iter().enumerate() {
        let outlets = group.outlets_for(family);
        if outlets.is_empty() {
            continue;
        }
        let selected = form
            .get(&format!("group_{idx}"))
            .and_then(|name| outlets.get_key_value(name));
        let Some((name, &value)) = selected else {
            return action(
                StatusCode::BAD_REQUEST,
                false,
                format!("无效的出口选择：{}", group.title),
                family,
            );
        };
        selected_labels.push(name);
        mark_value |= value & group.mask;
    }

    let hours = form.get("hours").and_then(|h| h.parse::<u32>().ok());
    let Some(hours) = hours.filter(|h| state.cfg.time_limits.contains(h)) else {
        return action(
            StatusCode::BAD_REQUEST,
            false,
            "无效的时限选择".into(),
            family,
        );
    };

    let ip_str = ip.to_string();
    if let Err(e) = state.nft.delete_element(&ip_str, map_name).await {
        tracing::error!("error deleting rule for {ip}: {e:#}");
    }
    match state
        .nft
        .add_element(&ip_str, mark_value, hours, map_name)
        .await
    {
        Ok(()) => {
            let message = format!(
                "IPv{family} 已开通：「{}」，{}",
                selected_labels.join(" + "),
                duration_label(hours)
            );
            action(StatusCode::OK, true, message, family)
        }
        Err(e) => {
            tracing::error!("error adding rule for {ip}: {e:#}");
            action(
                StatusCode::INTERNAL_SERVER_ERROR,
                false,
                "设置网络出口失败".into(),
                family,
            )
        }
    }
}

async fn api_close(
    State(state): State<AppState>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
) -> impl IntoResponse {
    let ip = normalize_ip(addr.ip());
    let family = ip_family(ip);
    if let Some(map_name) = state.cfg.map_for(family) {
        match state.nft.delete_element(&ip.to_string(), map_name).await {
            Ok(()) => return action(StatusCode::OK, true, format!("IPv{family} 已重置"), family),
            Err(e) => tracing::error!("error deleting rule for {ip}: {e:#}"),
        }
    }
    action(
        StatusCode::INTERNAL_SERVER_ERROR,
        false,
        "重置网络失败".into(),
        family,
    )
}

/// Allow cross-origin requests from configured domains and their subdomains
/// (the SPA on one split-horizon host fetches the sibling family's API; no
/// credentials).
fn origin_host_allowed(host: &str, domains: &[String]) -> bool {
    domains
        .iter()
        .any(|domain| host == domain || host.ends_with(&format!(".{domain}")))
}

fn cors_layer(domains: Vec<String>) -> CorsLayer {
    CorsLayer::new()
        .allow_origin(AllowOrigin::predicate(move |origin: &HeaderValue, _| {
            let Ok(origin) = origin.to_str() else {
                return false;
            };
            let host = origin
                .split_once("//")
                .map_or("", |(_, rest)| rest)
                .split(['/', ':'])
                .next()
                .unwrap_or("");
            origin_host_allowed(host, &domains)
        }))
        .allow_methods([axum::http::Method::GET, axum::http::Method::POST])
        .allow_headers([header::CONTENT_TYPE])
}

fn router(state: AppState) -> Router {
    let cors_domains = state.cfg.portal.cors_domains();
    let mut router = Router::new()
        .route("/", get(index))
        .route("/api/status", get(api_status))
        .route("/api/open", post(api_open))
        .route("/api/close", post(api_close))
        .with_state(state);
    if !cors_domains.is_empty() {
        router = router.layer(cors_layer(cors_domains));
    }
    router
}

async fn serve_http(router: Router, listen: String) -> Result<()> {
    let listener = tokio::net::TcpListener::bind(&listen)
        .await
        .with_context(|| format!("failed to bind {listen}"))?;
    tracing::info!("web listening on {listen}");
    axum::serve(
        listener,
        router.into_make_service_with_connect_info::<SocketAddr>(),
    )
    .await
    .context("web server failed")
}

async fn serve_https(router: Router, listen: String, tls: RustlsConfig) -> Result<()> {
    let addr: SocketAddr = listen
        .parse()
        .with_context(|| format!("invalid https listen address {listen}"))?;
    tracing::info!("web (https) listening on {addr}");
    axum_server::bind_rustls(addr, tls)
        .serve(router.into_make_service_with_connect_info::<SocketAddr>())
        .await
        .context("https server failed")
}

pub async fn serve(state: AppState, cfg: WebConfig) -> Result<()> {
    let router = router(state);
    let mut listeners: JoinSet<Result<()>> = JoinSet::new();
    for listen in cfg.listen {
        listeners.spawn(serve_http(router.clone(), listen));
    }
    if let Some(https) = cfg.https {
        let tls = RustlsConfig::from_pem_file(&https.cert, &https.key)
            .await
            .context("failed to load TLS certificate")?;
        for listen in https.listen {
            listeners.spawn(serve_https(router.clone(), listen, tls.clone()));
        }
    }
    // Listeners never exit on their own; treat the first return as fatal.
    match listeners.join_next().await {
        Some(result) => result.context("web listener panicked")?,
        None => Ok(()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::{PortalConfig, PortalHosts};
    use indexmap::IndexMap;

    #[test]
    fn renders_per_page_api_hosts_as_json() {
        let mut hosts = IndexMap::new();
        hosts.insert(
            "wlt.example.org".to_owned(),
            PortalHosts {
                v4_host: Some("wlt-ipv4.example.org".to_owned()),
                v6_host: Some("wlt-ipv6.example.org".to_owned()),
            },
        );
        hosts.insert(
            "wlt.example.net".to_owned(),
            PortalHosts {
                v4_host: None,
                v6_host: Some("wlt-ipv6.example.net".to_owned()),
            },
        );
        let portal = PortalConfig {
            v4_host: Some("wlt-ipv4.default.example".to_owned()),
            v6_host: Some("wlt-ipv6.default.example".to_owned()),
            hosts,
            cors_domain: None,
            cors_domains: vec![],
        };

        let data: serde_json::Value = serde_json::from_str(&api_hosts_json(&portal)).unwrap();

        assert_eq!(data["wlt.example.org"]["4"], "wlt-ipv4.example.org");
        assert_eq!(data["wlt.example.org"]["6"], "wlt-ipv6.example.org");
        assert_eq!(data["wlt.example.net"]["4"], "wlt-ipv4.default.example");
        assert_eq!(data["wlt.example.net"]["6"], "wlt-ipv6.example.net");
    }

    #[test]
    fn matches_exact_cors_domains_and_subdomains() {
        let domains = vec!["example.net".to_owned(), "example.org".to_owned()];

        assert!(origin_host_allowed("example.net", &domains));
        assert!(origin_host_allowed("wlt.example.net", &domains));
        assert!(origin_host_allowed("wlt.example.org", &domains));
        assert!(!origin_host_allowed("badexample.net", &domains));
        assert!(!origin_host_allowed("example.com", &domains));
    }
}
